internal/ssa2ast/func.go

```diff
@@ -0,0 +1,1161 @@
+					Tok: token.ASSIGN,
+					Rhs: []ast.Expr{&ast.TypeAssertExpr{X: valExpr, Type: assertTypeExpr}},
+				}
+			} else {
+				stmt = defineVar(i, &ast.TypeAssertExpr{X: valExpr, Type: assertTypeExpr})
+			}
+		case *ssa.UnOp:
+			valExpr, err := fc.convertSsaValue(i.X)
+			if err != nil {
+				return err
+			}
+			if i.CommaOk {
+				if i.Op != token.ARROW {
+					return fmt.Errorf("unary operator %s in %v: %w", i.Op, instr, ErrUnsupported)
+				}
+
+				valName, valType, valHasRefs := fc.tupleVarNameAndType(i, 0)
+				okName, okType, okHasRefs := fc.tupleVarNameAndType(i, 1)
+				if valHasRefs {
+					astFunc.Vars[valName] = valType
+				}
+				if okHasRefs {
+					astFunc.Vars[okName] = okType
+				}
+
+				stmt = &ast.AssignStmt{
+					Lhs: []ast.Expr{ast.NewIdent(valName), ast.NewIdent(okName)},
+					Tok: token.ASSIGN,
+					Rhs: []ast.Expr{&ast.UnaryExpr{
+						Op: token.ARROW,
+						X:  valExpr,
+					}},
+				}
+			} else if i.Op == token.MUL {
+				stmt = defineVar(i, &ast.StarExpr{X: valExpr})
+			} else {
+				stmt = defineVar(i, &ast.UnaryExpr{Op: i.Op, X: valExpr})
+			}
+		case *ssa.MakeClosure:
+			anonFunc := i.Fn.(*ssa.Function)
+			anonFuncName, err := fc.getAnonFunctionName(anonFunc)
+			if err != nil {
+				return err
+			}
+			if anonFuncName == nil {
+				return fmt.Errorf("make closure for non anon func %s: %w", anonFunc.Name(), ErrUnsupported)
+			}
+
+			callExpr := &ast.CallExpr{Fun: anonFuncName}
+			for _, freeVar := range i.Bindings {
+				varExr, err := fc.convertSsaValue(freeVar)
+				if err != nil {
+					return err
+				}
+				callExpr.Args = append(callExpr.Args, varExr)
+			}
+
+			stmt = defineVar(i, callExpr)
+		case *ssa.RunDefers, *ssa.DebugRef:
+			// ignored
+			continue
+		default:
+			return fmt.Errorf("instruction %v: %w", instr, ErrUnsupported)
+		}
+
+		if stmt != nil {
+			astBlock.Body = append(astBlock.Body, stmt)
+		}
+	}
+
+	exitInstr := ssaBlock.Instrs[len(ssaBlock.Instrs)-1]
+	switch exit := exitInstr.(type) {
+	case *ssa.Jump:
+		targetBlockIdx := ssaBlock.Succs[0].Index
+		astBlock.Exit = fc.gotoStmt(targetBlockIdx)
+	case *ssa.If:
+		tblock := ssaBlock.Succs[0].Index
+		fblock := ssaBlock.Succs[1].Index
+
+		condExpr, err := fc.convertSsaValue(exit.Cond)
+		if err != nil {
+			return err
+		}
+
+		astBlock.Exit = &ast.IfStmt{
+			Cond: condExpr,
+			Body: ah.BlockStmt(fc.gotoStmt(tblock)),
+			Else: ah.BlockStmt(fc.gotoStmt(fblock)),
+		}
+	case *ssa.Return:
+		exitStmt := &ast.ReturnStmt{}
+		for _, result := range exit.Results {
+			resultExpr, err := fc.convertSsaValue(result)
+			if err != nil {
+				return err
+			}
+			exitStmt.Results = append(exitStmt.Results, resultExpr)
+		}
+		astBlock.Exit = exitStmt
+	case *ssa.Panic:
+		panicArgExpr, err := fc.convertSsaValue(exit.X)
+		if err != nil {
+			return err
+		}
+		astBlock.Exit = &ast.ExprStmt{X: ah.CallExprByName("panic", panicArgExpr)}
+	default:
+		return fmt.Errorf("exit instruction %v: %w", exit, ErrUnsupported)
+	}
```

Comment:
panic?

Reply:
i don't think this is a good idea, ssa2ast is written in a library-like way and i don't see the point in changing it